src/iosource/Manager.cc

```diff
@@ -59,124 +109,411 @@ IOSource* Manager::FindSoonest(double* ts)
+			// TODO: does this make sense?
+			if ( events[i].data.fd == timerfd && events[i].events == EPOLLIN )
+				{
+				uint64_t elapsed;
+				read(timerfd, &elapsed, 8);
+
+				ready.clear();
+				if ( timeout_src )
+					ready.insert(timeout_src);
+				else
+					timer_expired = true;
+				break;
+				}
+			else
+				{
+				auto entry = fd_map.find(events[i].data.fd);
+				if ( entry != fd_map.end() )
+					{
+					if ( events[i].events == EPOLLIN )
+						ready.insert(entry->second);
+					else if ( events[i].events == EPOLLERR || events[i].events == EPOLLHUP )
+						printf("Source %s returned an error from poll (0x%x)\n",
+							entry->second->Tag(), events[i].events);
+					}
+				}
+			}
+		}
 
-	timeout.tv_sec = 0;
-	timeout.tv_usec = 0;
+	return timer_expired;
+	}
 
-	if ( select(maxx + 1, &fd_read, &fd_write, &fd_except, &timeout) > 0 )
-		{ // Find soonest.
-		for ( SourceList::iterator i = sources.begin();
-		      i != sources.end(); ++i )
+#elif defined(HAVE_KQUEUE)
+
+void Manager::InitQueue()
+	{
+	event_queue = kqueue();
+	if ( event_queue == -1 )
+		reporter->FatalError("Failed to initialize kqueue: %s", strerror(errno));
+	}
+
+void Manager::RegisterFd(int fd, IOSource* src)
+	{
+	struct kevent event;
+	EV_SET(&event, fd, EVFILT_READ, EV_ADD | EV_CLEAR, 0, 0, NULL);
+	int ret = kevent(event_queue, &event, 1, NULL, 0, NULL);
+	if ( ret != -1 )
+		{
+		events.push_back({0, 0, 0, 0, 0, NULL});
+		DBG_LOG(DBG_MAINLOOP, "Registered fd %d from %s", fd, src->Tag());
+		fd_map[fd] = src;
+
+		Wakeup("RegisterFd");
+		}
+	else
+		{
+		DBG_LOG(DBG_MAINLOOP, "Failed to register fd %d from %s: %s", fd, src->Tag(), strerror(errno));
+		}
+	}
+
+void Manager::UnregisterFd(int fd)
+	{
+	if ( fd_map.find(fd) != fd_map.end() )
+		{
+		struct kevent event;
+		EV_SET(&event, fd, EVFILT_READ, EV_DELETE, 0, 0, NULL);
+		int ret = kevent(event_queue, &event, 1, NULL, 0, NULL);
+		if ( ret != -1 )
+			DBG_LOG(DBG_MAINLOOP, "Unregistered fd %d", fd);
+
+		fd_map.erase(fd);
+
+		Wakeup("UnregisterFd");
+		}
+	}
+
+bool Manager::Poll(std::set<IOSource*>& ready, double timeout, IOSource* timeout_src)
+	{
+	bool timer_expired = false;
+
+	struct timespec kqueue_timeout;
+	ConvertTimeout(timeout, kqueue_timeout);
+
+	int ret = kevent(event_queue, NULL, 0, events.data(), events.size(), &kqueue_timeout);
+	if ( ret == -1 )
+		{
+		// Ignore interrupts since we may catch one during shutdown and we don't want the
+		// error to get printed.
+		if ( errno != EINTR )
+			perror("kevent error");
+		}
+	else if ( ret == 0 )
+		{
+		if ( timeout_src )
+			ready.insert(timeout_src);
+		else
+			// Setting this to true here but not returning a source causes net_run
+			// to advance the time and process timers as normal.
+			timer_expired = true;
+		}
+	else
+		{
+		// kevent returns the number of events that are ready, so we only need to loop
+		// over that many of them.
+		for ( int i = 0; i < ret; i++ )
```

Comment:
This doesn't have the additional timeout_src logic from above's EPOLL version (the TODO there). Does it not need to? Generally, it's good to minimize any logic inside these 3 #ifdefs that's actually independent of the polling implementation, to avoid risking getting into different behavior per case. So, in this case: if we need the additional logic for checking if timeout_src fired in all cases, can we move that into the code that's calling `Poll()`?

Reply:
The `kqueue` version doesn't have the extra logic because it can just take a `timespec` struct as a timeout value directly. The `epoll` and `poll` versions have to use a `timerfd` so they have to check whether that was what woke up the poll instead of being able to just check for what the return value is. It's definitely more ugly, but it's also a "feature" of epoll and poll to only take millisecond timeout values.